README.md

```diff
@@ -10,12 +10,328 @@ A core widget library for Dojo 2.
+
+### Base Widget
+
+The class `createWidgetBase` provides all base Dojo 2 widget functionality including caching and widget lifecycle management. It can be used directly or extended to create custom widgets.
+
+To customise the widget an optional `options` argument can be provided with the following interface.
+
+**Type**: `WidgetOptions<WidgetState>` - All properties are optional.
+
+|Property|Type|Description|
+|---|---|---|
+|id|string|identifier for the widget|
+|state|WidgetState|Initial state of the widget|
+|stateFrom|StoreObservablePatchable|Observable that provides state for the widget|
+|listeners|EventedListenersMap|Map of listeners for to attach to the widget|
+|tagName|string|Override the widgets     tagname|
+|getChildrenNodes|Function|Function that returns an array of children DNodes|
+|nodeAttributes|Function[]|An array of functions that return VNodeProperties to be applied to the VNode|
+
+By default the base widget class applies `id`, `classes` and `styles` from the widgets specified `state` (either by direct state injection or via an observable store).
+
+#### Simple Widgets
+To create a basic widget `createWidgetBase` can be used directly by importing the class.
+
+```ts
+import createWidgetBase from 'dojo-widgets/bases/createWidgetBase';
+
+const myBasicWidget = createWidgetBase();
+```
+
+Creates the following DOM element:
+
+```html
+<div></div>
+```
+The following example demonstrates how `id`, `classes` and `styles` are applied to the generated DOM.
+
+```ts
+import createWidgetBase from 'dojo-widgets/bases/createWidgetBase';
+
+const myBasicWidget = createWidgetBase({
+    state: {
+        id: 'my-widget',
+        classes: [ 'class-a', 'class-b' ],
+        styles: [ 'width:20px' ]
+    }
+});
+```
+
+Creates the following DOM element:
+
+```html
+<div data-widget-id="my-widget" class="class-a class-b" styles="width:20px"></div>
+```
+Alternatively state can be derived directly from an observable store passed via `stateFrom`. The following code will create the same DOM element as the above example.
+
+```ts
+const widgetStore = createObservableStore({
+    data: [
+        {
+            id: 'my-widget',
+            classes: [ 'class-a', 'class-b' ],
+            styles: [ 'width:20px' ]
+        }
+    ]
+});
+
+const myBasicWidget = createWidgetBase({
+    id: 'my-widget',
+    stateFrom: widgetStore
+});
+```
+
+#### `d`
+
+`d` is a function that is used within Dojo 2 to express widget hierarchical structure using both Dojo 2 widget factories or Hyperscript, it is imported via
+
+```ts
+import d from 'dojo-widgets/util/d';
+```
+
+and the param and return interfaces are available in the `dojo-interfaces` package.
+
+```ts
+import { DNode, HNode, WNode } from 'dojo-interfaces/widgetBases';
+```
+
+The API for using Hyperscript provides multiple signatures for convenience, **tagName** is the only mandatory argument, **options** is defaulted to `{}` when not provided and **children** is completely optional
```

Comment:
Maybe a link to Hyperscript? Maybe **options** defaults to ... in order to keep consistent tense?